src/main/java/im/status/keycard/SecureChannel.java

```diff
@@ -245,24 +349,24 @@ private short pairStep1(byte[] apduBuffer) {
    * @return the length of the reply
    */
   private short pairStep2(byte[] apduBuffer) {
-    crypto.sha256.update(pairingSecret, (short) 0, SC_SECRET_LENGTH);
-    crypto.sha256.doFinal(secret, (short) 0, SC_SECRET_LENGTH, secret, (short) 0);
-
+    // Compare client cryptogram to the expected cryptogram (stored in secret buffer)
     if (Util.arrayCompare(apduBuffer, ISO7816.OFFSET_CDATA, secret, (short) 0, SC_SECRET_LENGTH) != 0) {
       preassignedPairingOffset = -1;
       ISOException.throwIt(ISO7816.SW_SECURITY_STATUS_NOT_SATISFIED);
     }
 
+    // Generate random pairing salt & save pairing key. Copy pairing index and pairing salt into response buffer
+    // pairingKey = sha256(pairingSalt, cryptogram)
     crypto.random.generateData(apduBuffer, (short) 1, SC_SECRET_LENGTH);
-    crypto.sha256.update(pairingSecret, (short) 0, SC_SECRET_LENGTH);
-    crypto.sha256.doFinal(apduBuffer, (short) 1, SC_SECRET_LENGTH, pairingKeys, (short) (preassignedPairingOffset + 1));
+    crypto.sha256.update(apduBuffer, (short) 1, SC_SECRET_LENGTH);
```

Comment:
Woah what is you doin?

The order matters here. You want your output to go into `apduBuffer` right?

I have no idea why you would have willingly swapped these operations, but if that's the case, at least don't offset your `apduBuffer` by 1 here -- that byte was meant to store the pairing index

- [ ] Fix this or tell me I'm wrong

Reply:
The logic has changed here a bit, because the pairStep2 challenge verification is different, because the cryptogram secret is different, because pairStep1 no longer uses cryptogram. Basically, the pairingKey derivation here is different, because the preconditions have changed.

**Old logic:**
1) Generate random pairing salt (in APDU buffer to send back as response)
2) Derive pairing key as `sha256(pairingSecret | pairingSalt)`. Save the key, but don't send it!
3) Copy pairing index to response buffer

**New logic:** (ie, without `pairingSecret`)
1) Generate random pairing salt (in APDU buffer to send back as response)
2) Derive pairing key as `sha256(pairingSalt | pairingCryptogram)` // note, we don't use pairingSecret anymore
3) Copy pairing index to response buffer

Reply:
Sorry.. I think I misread something when I reviewed this 🙃